rfcs/0091-pure-object-prototypes.md

```diff
@@ -0,0 +1,418 @@
+  See [RFC 0082 lib.experimental](https://github.com/NixOS/rfcs/pull/82).
+
+- Performance:
+  Computing the `precedenceList` for an object in general is
+  linear in the size of the inheritance graph (nodes plus arcs),
+  i.e. the number of transitive `supers` entries.
+  In the worst case, that's up to quadratic for each new object,
+  which is cubic in the total number of objects.
+  That can be slow.
+  That said, the same "computation" would have to be done by hand
+  by users who would want to achieve the same effect without automation,
+  so it's not really a drawback *given the desired effect*.
+
+- Missing features? Previous extension systems may have important features
+  that I have neglected so far, and that would need to be implemented
+  before we have "the" satisfactory object system for Nix.
+  For instance, magic argument processing from pkgs or some other scope;
+  and whatever feature modules need.
+  That would be one more reason to put the object in experimental until stable.
+  But the current state of POP as well as other copies with more features in progress
+  could all be in experimental until the ultimate object system wins.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Embrace non-modularity
+
+Keep the existing menagerie of extension systems, embrace the non-modularity of it.
+
+## Invent an even better object system
+
+We can invent an even better object system and still include this one in nixpkgs in the meantime.
+
+## Implement objects at a deeper level
+
+We can make objects part of the language at a deeper level, as in Jsonnet,
+for better performance and/or better error messages.
+But if we can afford a user-level implementation, that is more flexible, and
+we don't deep magic to fix or extend the object system.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+If we could agree on what "the" default prototype representation were for Nix,
+or autodetect which it is, we could likewise automatically wrap a traditional extension into a POP.
+Actually, we could probably autodetect whether an attrset has some `__unfix__`, `extend`
+or some such field. When passed an extension function, we could pass it the `self` argument,
+if it returns an attrset make that the attrset-to-merge, and if it returns another function,
+pass it the super and use the result as the attrset-to-merge.
+I'm not sure if backward-compatibility automagic is an asset or a liability,
+so I left it out for now. But if at some point the goal is to replace the existing zoo
+with a single improved solution, then we will want explicit conversion,
+if not implicit conversion.
+
+# Future work
+[future]: #future-work
+
+- Use and enjoy POP in more subsystems of nixpkgs than Gerbil.
+- See how POP, with or without further improvements,
+  can or cannot fully replace other Nix extension systems.
+- In particular, see how POP, with or without further improvements,
+  may improve the situation for modules.
```

Comment:
I would classify comparison with the current module system as an unresolved question…

Reply:
I added a paragraph there, too.